specutils/core/generic.py

```diff
@@ -0,0 +1,124 @@
+from specutils.spectrum1d import Spectrum1D
+
+import numpy as np
+import logging
+
+from astropy.nddata import NDArithmeticMixin, NDSlicingMixin, NDIOMixin
+import astropy.units as u
+from astropy.wcs import WCS
+
+
+class GenericSpectrum1D(NDIOMixin, NDSlicingMixin, NDArithmeticMixin,
+                        Spectrum1D):
+    """
+    A generalized spectrum data object that implements several `NDData`
+    features such as IO, arithmetic, and slicing.
+    """
+
+    # The Spectrum1D wcs_attribute object is useless to us, let's avoid it
+    _wcs_attributes = {}
+
+    def __init__(self, data, name="", dispersion=None, dispersion_unit=None,
+                 uncertainty=None, wcs=None, *args, **kwargs):
+        super(GenericSpectrum1D, self).__init__(flux=data,
+                                                wcs=wcs,
+                                                uncertainty=uncertainty,
+                                                indexer="UNUSED",
+                                                *args, **kwargs)
+        self._dispersion = dispersion
+        self._dispersion_unit = dispersion_unit
+        self.name = name or "New Data Object"
+
+    @classmethod
+    def copy(cls, original, deep_copy=True, **kwargs):
+        """
+        Create a new `GenericSpectrum1D` object using current property
+        values. Takes all the arguments a Spectrum1D expects, arguments that
+        are not included use this instance's values.
+        """
+        self_kwargs = {"data": original._data,
+                       "dispersion": original._dispersion,
+                       "unit": original.unit, "wcs": original.wcs,
+                       "uncertainty": original._uncertainty,
+                       "mask": original.mask, "meta": original.meta}
+
+        self_kwargs.update(kwargs)
+
+        return cls(copy=deep_copy, **self_kwargs)
+
+    @classmethod
+    def from_array(cls, data, *args, **kwargs):
+        return cls(name="Array Data", data=data, *args, **kwargs)
+
+    @property
+    def dispersion(self):
+        """
+        Override parent method in order to construct dispersion array from
+        Astropy's WCS object instead of the custom specutils WCS object.
+        """
+        if self._dispersion is None:
+            self._dispersion = np.arange(self.data.size)
+
+            if isinstance(self.wcs, WCS):
+                try:
+                    crval = self.wcs.wcs.crval[0]
```

Comment:
> Or alternatively, why not use the machinery described here:

I don't see the point in this, really. Either we get the relevant information from the wcs, or directly from the fits header. In both cases (theirs and ours), we're doing essentially the same thing, but just that we're not incorporating an object dependency that would otherwise never be used in our case (e.g. `specwcs.Spectrum1DPolynomialWCS`).

But, using `all_pix2world` is definitely an improvement, please let me know if the changes are satisfactory.